### PR TITLE
[SIP-15A] Presto types (PoC)

### PR DESCRIPTION
**Pull request by john-bodley**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

This is very much a Proof-of-Concept (PoC) but I though there was merit in sharing early my thoughts as to how I would implement aspects of [SIP-15A](https://github.com/apache/incubator-superset/issues/7656) in relation to how we could think about type conversions in the future. 

The premise is Superset needs to transform either Python types (such as `datetime`) or either a SQL type (`DATE`, `TIMESTAMP`, etc.), Python type (`str`, `int`, `float`, etc.), or time grain (`P1D`, `P1W` etc). Rather than depending on methods like `convert_dttm`, `epoch_to_dttm`, etc. I felt it would make more sense to define an engine specific graph where the nodes are the various temporal types and the edges represents the SQL expression required to convert between the types. A shortest path algorithm is used to define the path (and thus SQL) between the types.

I though the merits are:
- Helps to ensure all the mappings reside in the same place. Note it's still somewhat fragmented due the constraints on the existing API. 
- Removes redundancy, i.e., in Hive there's a bug in the `from_unixtime` UDF which actually requires that the unix time be a float rather than an integer and thus one would define two edges with minimal SQL, i.e.,

```
graph = nx.DiGraph()
graph.add_edge(BigInteger, Float, sql='CAST({col} AS DOUBLE)')
graph.add_edge(Float, TIMESTAMP, sql='CAST({col} AS TIMESTAMP)')
```
    
which would provide these mappings; (`BigInteger`, `TIMESTAMP`) and (`Float`, `TIMESTAMP`).

Note this PR does not address several aspects of SIP-15A including type inferencing and enforcing ISO 8601 encoding. Additionally if rolled out there would be additional major refactoring of the`BaseEngineSpec` and `TableColumn` classes.

### TEST PLAN

N/A

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

to: @agrawaldevesh @betodealmeida @michellethomas @mistercrunch @villebro

### CODE REVIEW

**Comment by villebro** [comment]
This sounds neat, as I tend to get confused by all the different conversion functions. I will take a closer look over the weekend. Btw, not that it matters for testing, but there's a conflict that needs rebasing.

File: superset/db_engine_specs/presto.py [hunk 1]
```diff
@@ -38,25 +44,33 @@
 
 
 class PrestoEngineSpec(BaseEngineSpec):
+    import pyhive.sqlalchemy_presto
+
     engine = 'presto'
+    type_map = pyhive.sqlalchemy_presto._type_map
 
     time_grain_functions = {
```
Comment (john-bodley) [inline superset/db_engine_specs/presto.py]: This logic would probably be refactored if we rolled this our globally.

File: superset/db_engine_specs/presto.py [hunk 2]
```diff
@@ -349,13 +363,127 @@ def adjust_database_uri(cls, uri, selected_schema=None):
         return uri
 
     @classmethod
-    def convert_dttm(cls, target_type, dttm):
```
Comment (john-bodley) [inline superset/db_engine_specs/presto.py]: This logic would probably be refactored if we rolled this our globally.

File: superset/connectors/sqla/models.py [hunk 3]
```diff
@@ -190,11 +190,21 @@ def dttm_sql_literal(self, dttm):
...
         else:
             s = self.table.database.db_engine_spec.convert_dttm(
                 self.type or '', dttm)
-            return s or "'{}'".format(dttm.strftime('%Y-%m-%d %H:%M:%S.%f'))
+
+            if s:
+                return s
+
+            if app.config['SIP_15_ENABLED']:
+                raise TypeError(f"The '{type}' type is unsupported.")
```
Comment (villebro) [inline superset/connectors/sqla/models.py]: shouldn't this be `{self.type}`?

**Review by villebro — APPROVED** [review]
Very exciting 👍 LGTM apart from possible typo. Is the migration plan to first introduce this in Presto, and once stability has been established through real-life testing, move stuff from `PrestoEngineSpec` to `BaseEngineSpec` and implement accordingly in all other `*Spec`s?

File: superset/db_engine_specs/presto.py [hunk 4]
```diff
@@ -38,26 +44,34 @@
 
 
 class PrestoEngineSpec(BaseEngineSpec):
+    import pyhive.sqlalchemy_presto
```
Comment (mistercrunch) [inline superset/db_engine_specs/presto.py]: I think this is essentially the same as importing in module scope, it'll fail if `pyhive` isn't installed

**Pull request closed without merging**